File: backend/routes/hotels.py
"""
TravelVista — Hotels Routes
"""
import time
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import get_db, hotel_to_dict

hotels_bp = Blueprint('hotels', __name__, url_prefix='/api/hotels')


def _is_admin(uid):
    conn = get_db()
    row = conn.execute("SELECT role FROM users WHERE id = ?", (uid,)).fetchone()
    conn.close()
    return row and row['role'] == 'admin'


@hotels_bp.route('', methods=['GET'])
@jwt_required()
def list_hotels():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute("SELECT * FROM hotels").fetchall()
    conn.close()
    return jsonify([hotel_to_dict(r) for r in rows])


@hotels_bp.route('', methods=['POST'])
@jwt_required()
def add_hotel():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    htl_id = 'htl-' + str(int(time.time() * 1000))

    conn = get_db()
    conn.execute(
        "INSERT INTO hotels (id,name,location,rating,price_per_night,capacity,type,image) VALUES (?,?,?,?,?,?,?,?)",
        (htl_id, data.get('name', ''), data.get('location', ''),
         data.get('rating', 0), data.get('pricePerNight', 0),
         data.get('capacity', 0), data.get('type', 'Hotel'), data.get('image', ''))
    )
    conn.commit()
    row = conn.execute("SELECT * FROM hotels WHERE id = ?", (htl_id,)).fetchone()
    conn.close()
    return jsonify(hotel_to_dict(row)), 201


@hotels_bp.route('/<htl_id>', methods=['PUT'])
@jwt_required()
def update_hotel(htl_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    conn = get_db()
    conn.execute(
        """UPDATE hotels SET name=?, location=?, rating=?, price_per_night=?,
           capacity=?, type=?, image=? WHERE id=?""",
        (data.get('name', ''), data.get('location', ''),
         data.get('rating', 0), data.get('pricePerNight', 0),
         data.get('capacity', 0), data.get('type', 'Hotel'),
         data.get('image', ''), htl_id)
    )
    conn.commit()
    row = conn.execute("SELECT * FROM hotels WHERE id = ?", (htl_id,)).fetchone()
    conn.close()
    if not row:
        return jsonify({'message': 'Hotel not found'}), 404
    return jsonify(hotel_to_dict(row))


@hotels_bp.route('/<htl_id>', methods=['DELETE'])
@jwt_required()
def delete_hotel(htl_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    conn.execute("DELETE FROM hotels WHERE id = ?", (htl_id,))
    conn.commit()
    conn.close()
    return jsonify({'success': True})

File: api/hotels.php
<?php
/**
 * TravelVista — Hotels API (Admin)
 * GET                — List hotels
 * POST               — Add hotel
 * PUT    ?id=X       — Update hotel
 * DELETE ?id=X       — Delete hotel
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') { http_response_code(200); exit; }

require_once __DIR__ . '/../config/jwt_helper.php';

$method = $_SERVER['REQUEST_METHOD'];
$id = $_GET['id'] ?? '';
if (!$id) {
    $uri = $_SERVER['REQUEST_URI'] ?? '';
    if (preg_match('#/api/hotels/(\d+)#', $uri, $m)) { $id = $m[1]; }
}

// ── GET: List hotels ──
if ($method === 'GET') {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT * FROM hotels")->fetchAll();
    json_response(array_map('hotel_to_dict', $rows));
}

// ── POST: Add hotel ──
if ($method === 'POST') {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare(
        "INSERT INTO hotels (name, location, rating, price_per_night, capacity, type, image) VALUES (?, ?, ?, ?, ?, ?, ?)"
    );
    $stmt->execute([
        $data['name'] ?? '', $data['location'] ?? '',
        $data['rating'] ?? 0, $data['pricePerNight'] ?? 0,
        $data['capacity'] ?? 0, $data['type'] ?? 'Hotel', $data['image'] ?? '',
    ]);

    $htlId = $db->lastInsertId();
    $stmt = $db->prepare("SELECT * FROM hotels WHERE id = ?");
    $stmt->execute([$htlId]);
    json_response(hotel_to_dict($stmt->fetch()), 201);
}

// ── PUT: Update hotel ──
if ($method === 'PUT' && $id) {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $db->prepare(
        "UPDATE hotels SET name=?, location=?, rating=?, price_per_night=?, capacity=?, type=?, image=? WHERE id=?"
    )->execute([
        $data['name'] ?? '', $data['location'] ?? '',
        $data['rating'] ?? 0, $data['pricePerNight'] ?? 0,
        $data['capacity'] ?? 0, $data['type'] ?? 'Hotel',
        $data['image'] ?? '', $id,
    ]);

    $stmt = $db->prepare("SELECT * FROM hotels WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Hotel not found'], 404);
    json_response(hotel_to_dict($row));
}

// ── DELETE: Delete hotel ──
if ($method === 'DELETE' && $id) {
    require_admin();
    $db = getDB();
    $db->prepare("DELETE FROM hotels WHERE id = ?")->execute([$id]);
    json_response(['success' => true]);
}

json_response(['message' => 'Invalid request'], 400);
?>
